mpi3501.cpp

```diff
@@ -1,135 +1,152 @@
+  {
+    CLEAR_GPIO(GPIO_SPI0_CE0); // Enable Touch
+    CLEAR_GPIO(GPIO_SPI0_CE1); // Enable Display
+
+    BEGIN_SPI_COMMUNICATION();
+
+    usleep(25*1000);
+
+    SET_GPIO(GPIO_SPI0_CE0); // Disable Touch
+    usleep(25*1000);
+
+    SPI_TRANSFER(0x00001100); // Oy!  Now this is a reset
+    usleep(10*1000);
+    SPI_TRANSFER(0xff001100);
+    SPI_TRANSFER(0xff001100);
+    usleep(10*1000);
+    SPI_TRANSFER(0xff001100);
+    SPI_TRANSFER(0xff001100);
+    SPI_TRANSFER(0xff001100);
+    SPI_TRANSFER(0xff001100);
+    usleep(15*1000);
+    SPI_TRANSFER(0x11001100/*Sleep Out*/);
+    usleep(150*1000);
+
+    SPI_TRANSFER(0xB0001100, 0x00, 0x00); // CLK 30Hz?
+    SPI_TRANSFER(0xB3001100, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00); // set RGB?
+    SPI_TRANSFER(0xB9001100, 0x00, 0x01, 0x00, 0x00, 0x00, 0x0F, 0x00, 0x0f); // ext cmd?
+    SPI_TRANSFER(0xC0001100, 0x00, 0x13, 0x00, 0x3B, 0x00, 0x00, 0x00, 0x02
+	, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x43);                    // Timing?
+    SPI_TRANSFER(0xC1001100, 0x00, 0x08, 0x00, 0x0F, 0x00, 0x08, 0x00, 0x08); // gamma?
+    SPI_TRANSFER(0xC4001100, 0x00, 0x11, 0x00, 0x07, 0x00, 0x03, 0x00, 0x04); // sleep modes?
+    SPI_TRANSFER(0xC6001100, 0x00, 0x00); // ?
+    SPI_TRANSFER(0xC8001100, 0x00, 0x03, 0x00, 0x03, 0x00, 0x13, 0x00, 0x5C
+	, 0x00, 0x03, 0x00, 0x07, 0x00, 0x14, 0x00, 0x08
+	, 0x00, 0x00, 0x00, 0x21, 0x00, 0x08, 0x00, 0x14
+	, 0x00, 0x07, 0x00, 0x53, 0x00, 0x0C, 0x00, 0x13
+	, 0x00, 0x03, 0x00, 0x03, 0x00, 0x21, 0x00, 0x00); // ?
+    SPI_TRANSFER(0x35001100, 0x00, 0x00);
+
+#define MADCTL_BGR_PIXEL_ORDER (1<<3)
+#define MADCTL_ROW_COLUMN_EXCHANGE (1<<5)
+#define MADCTL_COLUMN_ADDRESS_ORDER_SWAP (1<<6)
+#define MADCTL_ROW_ADDRESS_ORDER_SWAP (1<<7)
+#define MADCTL_ROTATE_180_DEGREES (MADCTL_COLUMN_ADDRESS_ORDER_SWAP | MADCTL_ROW_ADDRESS_ORDER_SWAP)
+
+    uint8_t madctl = 0;
+#ifndef DISPLAY_SWAP_BGR
+    madctl |= MADCTL_BGR_PIXEL_ORDER;
+#endif
+#if defined(DISPLAY_FLIP_ORIENTATION_IN_HARDWARE)
+    madctl |= MADCTL_ROW_COLUMN_EXCHANGE;
+#endif
+#ifdef DISPLAY_ROTATE_180_DEGREES
+    madctl ^= MADCTL_ROTATE_180_DEGREES;
+#endif
+    //SPI_TRANSFER(0x36001100/*MADCTL: Memory Access Control*/, 0x00, madctl);
+    SPI_TRANSFER(0x36001100/*MADCTL: Memory Access Control*/, 0x00, 0x60);
+
+    SPI_TRANSFER(0x3A001100/*Interface Pixel Format*/, 0x00, 0x55);
+    SPI_TRANSFER(0x44001100, 0x00, 0x00, 0x00, 0x01);
+    SPI_TRANSFER(0xD0001100, 0x00, 0x07, 0x00, 0x07, 0x00, 0x1D, 0x00, 0x03); // ? 
+    SPI_TRANSFER(0xD1001100, 0x00, 0x03, 0x00, 0x30, 0x00, 0x10); // ? 
+    SPI_TRANSFER(0xD2001100, 0x00, 0x03, 0x00, 0x14, 0x00, 0x04); // ? 
+    SPI_TRANSFER(0x29001100/*Display ON*/);
+
+    usleep(30*1000);
+
+    SPI_TRANSFER(0x2A001100, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x3F); // ? 
+    SPI_TRANSFER(0x2B001100, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0xE0); // ? 
+    SPI_TRANSFER(0xB4001100, 0x00, 0x00);
+    SPI_TRANSFER(0x2C001100); // x-fer mcu to frame memory?
```

Comment:
The last line with 2C is redundant here, and can be removed. This is because fbcp-ili9341 issues 2C command before each write, so this is not needed.

The driver also routinely issues 2A and 2B commands, which set write window coordinates, so if B4 does not depend on 2A and 2B, then the lines with 2A and 2B can also be removed. Not sure what B4 does though.

Reply:
I'll try removing them.  Honestly, I'd transposed the code with only emerging understanding at near-verbatim to what worked.  Now that there is working reference, will adjust as per your suggestions.

Reply:
Removed all suggested lines and it works fine.